File: include/PSModelData.java
package edu.uwstout.berriers.PSHelper.Model;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

import javax.xml.stream.XMLStreamException;
import javax.xml.stream.XMLStreamReader;

public class PSModelData {

	private long mFaceCount, mVertexCount;
	private File mZipFile;
	private String mMeshFilepath;
	private boolean mHasVtxColors, mHasUV;

	private Map<Integer, String> textureFiles;

	public PSModelData(File pZipFile) {
		mZipFile = pZipFile;
		
		mFaceCount = mVertexCount = -1;
		mHasVtxColors = mHasUV = false;
		mMeshFilepath = "";
		
		textureFiles = new HashMap<Integer, String>();
	}
	
	public void setFaceCount(long pFaceCount) { mFaceCount = pFaceCount; }
	public void setVertexCount(long pVertexCount) { mVertexCount = pVertexCount; }
	public void setHasVertexColors(boolean pHasVtxColors) { mHasVtxColors = pHasVtxColors; }
	public void setHasUV(boolean pHasUV) { mHasUV = pHasUV; }
	public void setMeshFilename(String pMeshFilepath) { mMeshFilepath = pMeshFilepath; }
	
	public void addTextureFile(int pId, String pFilepath) {
		if(textureFiles.containsKey(pId)) {
			System.err.println("Warning: possible texture ID collision (" + pId + " already exists)");
		}
		
		textureFiles.put(pId, pFilepath);
	}

	public File getArchiveFile() { return mZipFile; }

	public long getFaceCount() { return mFaceCount; }
	public long getVertexCount() { return mVertexCount; }
	public String getMeshFilename() { return mMeshFilepath; }
	public boolean hasVtxColors() { return mHasVtxColors; }
	public boolean hasUV() { return mHasUV; }

	public Map<Integer, String> getTextureFiles() { return textureFiles; }
	public String getTextureFile(int id) { return textureFiles.get(id); }
	
	public static PSModelData makeFromXML(XMLStreamReader reader, File pZipFile, PSChunkData pParent) throws XMLStreamException {
		// Sanity check
		if(reader.getEventType() == XMLStreamReader.START_DOCUMENT) { reader.nextTag(); }
		if(reader == null || reader.getLocalName() != "model" || !reader.isStartElement()) {
			return null;
		}

		// Make a new object
		PSModelData newModel = new PSModelData(pZipFile);

		// Parse the remaining XML data
		try {
			while(reader.hasNext()) {
			    reader.next();
			    // Tag open cases
			    if(reader.isStartElement()) {
			    	switch(reader.getLocalName()) {
			    	
			    		// Data inside the 'mesh' tag
			    		case "hasVertexColors": newModel.mHasVtxColors = reader.getElementText().equalsIgnoreCase("true"); break;
			    		case "hasUV": newModel.mHasVtxColors = reader.getElementText().equalsIgnoreCase("true"); break;
			    		case "faceCount": newModel.mFaceCount = Long.parseLong(reader.getElementText()); break;
			    		case "vertexCount": newModel.mVertexCount = Long.parseLong(reader.getElementText()); break;

			    		// Start of the mesh tag
			    		case "mesh":
			    			newModel.mMeshFilepath = reader.getAttributeValue(null, "path");
			    		break;

			    		// Texture tag
			    		case "texture":
			    		{
			    			String texPath = reader.getAttributeValue(null, "path");

			    			// Note, sometimes textures don't have an ID (if there's only one texture for instance)
			    			try {
			    				int texID = Integer.parseInt(reader.getAttributeValue(null, "id"));
			    				newModel.addTextureFile(texID, texPath);
			    			} catch(Exception e) { newModel.addTextureFile(0, texPath); }
			    		}
			    		break;

			    		// From inside the meta tag
            			case "property":
            			{
                        	String lPropertyName = reader.getAttributeValue(null, "name");
                        	String lPropertyValue = reader.getAttributeValue(null, "value");

                        	// Sometimes, this is the only way to get the face count (on older file versions)
                        	if(lPropertyName.contains("face_count") && newModel.mFaceCount <= 0) {
                        		newModel.mFaceCount = Long.parseLong(lPropertyValue);
                        	}
                        	
                        	pParent.parseChunkProperty(lPropertyName, lPropertyValue);
            			}
            			break;
			    	}
			    }

			    // Tag close cases
			    if(reader.isEndElement()) {
			    	if(reader.getLocalName().equalsIgnoreCase("model")) {
			    		return newModel;
			    	}
			    }
			}
		} catch (XMLStreamException e) {
			throw e;
		}

		// Should never reach this except when XML is malformed
		return null;
	}	
}

File: src/Model/PSModelData.cpp
#include "PSModelData.h"

#include "PSChunkData.h"

#include <QXmlStreamReader>
#include <QFile>

PSModelData::PSModelData(QFile* pZipFile) {
    mZipFile = pZipFile;

    mFaceCount = mVertexCount = -1;
    mHasVtxColors = mHasUV = false;
    mMeshFilepath = "";
}

PSModelData::~PSModelData() { }

void PSModelData::setFaceCount(long pFaceCount) { mFaceCount = pFaceCount; }
void PSModelData::setVertexCount(long pVertexCount) { mVertexCount = pVertexCount; }
void PSModelData::setHasVertexColors(bool pHasVtxColors) { mHasVtxColors = pHasVtxColors; }
void PSModelData::setHasUV(bool pHasUV) { mHasUV = pHasUV; }
void PSModelData::setMeshFilename(QString pMeshFilepath) { mMeshFilepath = pMeshFilepath; }

void PSModelData::addTextureFile(int pId, QString pFilepath) {
    if(textureFiles.contains(pId)) {
        qWarning("Warning: possible texture ID collision (%d already exists)\n", pId);
    }

    textureFiles.insert(pId, pFilepath);
}

QFile* PSModelData::getArchiveFile() { return mZipFile; }

long PSModelData::getFaceCount() { return mFaceCount; }
long PSModelData::getVertexCount() { return mVertexCount; }
QString PSModelData::getMeshFilename() { return mMeshFilepath; }
bool PSModelData::hasVtxColors() { return mHasVtxColors; }
bool PSModelData::hasUV() { return mHasUV; }

QMap<int, QString> PSModelData::getTextureFiles() { return textureFiles; }
QString PSModelData::getTextureFile(int id) { return textureFiles.value(id); }

PSModelData* PSModelData::makeFromXML(QXmlStreamReader* reader, QFile* pZipFile, PSChunkData* pParent) {
    // If this is a fresh XML doc, push to first tag.
    if(reader->tokenType() == QXmlStreamReader::StartDocument) {
        reader->readNextStartElement();
    }

    // Sanity check
    if(reader == NULL || reader->name() != "model" || !reader->isStartElement()) {
        return NULL;
    }

    // Make a new object
    PSModelData* newModel = new PSModelData(pZipFile);

    // Parse the remaining XML data
    try {
        while(!reader->atEnd()) {
            reader->readNext();
            if(reader->isStartElement()) {
                // Data inside the 'mesh' tag
                if (reader->name() == "hasVertexColors") { newModel->mHasVtxColors = (reader->readElementText() == "true"); }
                else if (reader->name() == "hasUV") { newModel->mHasUV = (reader->readElementText() == "true"); }
                else if (reader->name() == "faceCount") { newModel->mFaceCount = reader->readElementText().toLong(); }
                else if (reader->name() == "vertexCount") { newModel->mVertexCount = reader->readElementText().toLong(); }

                // Start of the mesh tag
                else if (reader->name() == "mesh") { newModel->mMeshFilepath = reader->attributes().value("", "path").toString(); }

                // Texture tag
                else if (reader->name() == "texture") {
                    QString texPath = reader->attributes().value("", "path").toString();

                    // Note, sometimes textures don't have an ID (if there's only one texture for instance)
                    if (reader->attributes().hasAttribute("", "id")) {
                        int texID = reader->attributes().value("", "id").toInt();
                        newModel->addTextureFile(texID, texPath);
                    } else {
                        newModel->addTextureFile(0, texPath);
                    }
                }

                // From inside the meta tag
                else if (reader->name() == "property") {
                    QString lPropertyName = reader->attributes().value("", "name").toString();
                    QString lPropertyValue = reader->attributes().value("", "value").toString();

                    // Sometimes, this is the only way to get the face count (on older file versions)
                    if(lPropertyName.contains("face_count") && newModel->mFaceCount <= 0) {
                        newModel->mFaceCount = lPropertyValue.toLong();
                    }

                    pParent->parseChunkProperty(lPropertyName, lPropertyValue);
                }
            }

            // Tag close cases
            if(reader->isEndElement()) {
                if(reader->name() == "model") {
                    return newModel;
                }
            }
        }
    } catch (...) {
        throw new std::logic_error("Error parsing model XML tag to PSModelData");
    }

    // Should never reach this except when XML is malformed
    return NULL;
}
